### DISCUSSION TITLE
Uncaught exception in middleware

### DISCUSSION THREAD

**Discussion by DZ8540** [adonisjs/core#3496] [Help] [answered]
Hello there, i have a problem. I want to check user role, i do this in my service class, where i'm throwing an error. Middleware have try catch construction where i'm use this service.

<img width="820" alt="image" src="https://user-images.githubusercontent.com/50152205/150117570-f8a9b8a6-1bb2-4e0d-913e-35e6603a9d9a.png">

and my service

<img width="578" alt="image" src="https://user-images.githubusercontent.com/50152205/150117642-3617a922-9ee6-4d8b-b65d-c447c3bc4769.png">

But my process will stop in the console, and the server will stop. What can I do to keep my server from stopping?

<img width="846" alt="image" src="https://user-images.githubusercontent.com/50152205/150118173-12fd6163-ea87-4e74-a3b0-63d747e4d033.png">

**Comment by DZ8540** [comment]
And can i get role properties without model's load method?

**Comment by tomgobich** [comment, marked as answer]
If I had to guess I'd say it's because your `this.checkAdminRole` call in your middleware is missing `await`, since the method is async.

As for your second question, in order to get the role properties you'll either need to load it or query it separately. Typically, what I do is have a seed that explicitly sets the role's ids. Then, define an enum within my application called `Roles` with each role record. This way I can perform the check without needing to make an additional database call.

For example:
```typescript
// contracts/Enums/Roles.ts

enum Roles {
  ADMIN = 1,
  USER = 2
}

export default Roles
```

```typescript
// app/Services/RoleService.ts

import Roles from 'Contracts/Enums/Roles'

export default class RoleService {
  public static checkAdminRole(user: User) {
    return user.roleId === Roles.ADMIN
  }
}
```

**Reply by DZ8540** [reply]
Thanks for your answer, apparently I made a lot of work, since I did not see async in my service's method. The idea with the enum is very good, thanks, but the second question remains open, since in other cases of relationships, I don’t feel comfortable when i calling load method(

**Reply by tomgobich** [reply]
An alternative to using load is preload using the query builder. Unfortunately, one of these two will be needed in order to include relationship information onto a record. If you're coming from somewhere like Laravel or .NET where the relationship is lazy-loaded upon request, Aman Virk wrote somewhere why this isn't in AdonisJS, but I don't have where that was saved for reference. I believe it was because there wasn't a known efficient way to go about it within Node, but I may be misremembering that.

```typescript
const user = await User.query()
  .where('id', 1)
  .preload('role')
  .firstOrFail()

console.log(user.role.name)
```
https://docs.adonisjs.com/reference/orm/query-builder#preload

Alternatively, if for example, you just need the `role` and don't need it attached to the `user` you can query through the user to it's relations as shown here:
https://docs.adonisjs.com/reference/orm/base-model#related

**Reply by DZ8540** [reply]
Thank you very much)

**Reply by tomgobich** [reply]
Anytime!

**Reply by DZ8540** [reply]
Can be extended model or database query builder for get relationships in one method.
https://docs.adonisjs.com/guides/database/query-builder#extending-modelquerybuilder

**Reply by tomgobich** [reply]
Yeah, I don't see why not. I've never tried doing it before, but inside the macro `this` should be the query builder, so you can call preload inside the macro for whatever you need. In that example, they're awaiting the result of the query and returning back. So, if you want it to still be chainable after your extended method, don't await and return back `this`, I think. 

Something like this, again keep in mind I've never attempted this so I may have something wrong here.
```typescript
ModelQueryBuilder.macro('getRole', async function () {
  return this.preload('role')
})
```

You can also use Query Scopes to do this as well, though I don't think the usage is as clean as what you're trying to get.
https://docs.adonisjs.com/guides/models/query-scopes#document

**Reply by DZ8540** [reply]
Query scopes should be minimalistic and clean. But it's best to use a macro. I roughly did the same with the default pagination.

Example: 
<img width="960" alt="image" src="https://user-images.githubusercontent.com/50152205/150290680-d5d75274-3f9e-4e43-b7cd-47a68942b742.png">
